### PR TITLE
Implement shared message ownership for shmem transport

### PR DESCRIPTION
**Pull request by rbx**
Usage:
```C++
MessagePtr original(factory->CreateMessage(size));
// ...
MessagePtr copy(factory->CreateMessage());
copy->Copy(*original);
```
Also add tests and docs for it.

### CODE REVIEW

File: test/message/_message.cxx [hunk 1]
```diff
@@ -160,14 +164,22 @@ auto RunPushPullWithAlignment(string const& transport, string const& _address) -
...
+    size_t const size25{25};
+    size_t const size50{50};
+    Alignment const align64{64};
 
-        auto msg(push.NewMessage(size25));
-        msg->Rebuild(size50, align);
-        ASSERT_TRUE(CheckMsgAlignment(*msg, align));
-    }
+    auto msg1(push.NewMessage(size25));
+    msg1->Rebuild(size50, fair::mq::Alignment{align64});
```
Comment (dennisklein) [inline test/message/_message.cxx:172]:
```suggestion
    msg1->Rebuild(size50, align64);
```

Or am I missing something?

File: test/message/_message.cxx [hunk 2]
```diff
@@ -182,13 +194,164 @@ auto EmptyMessage(string const& transport, string const& _address) -> void
...
+    const size_t size = 2;
+    MessagePtr original(factory->CreateMessage(size));
+    memcpy(original->GetData(), "AB", size);
+    {
+        MessagePtr copy(factory->CreateMessage());
+        copy->Copy(*original);
+        EXPECT_EQ(original->GetSize(), copy->GetSize());
+        EXPECT_EQ(original->GetData(), copy->GetData());
+        EXPECT_EQ(static_cast<const shmem::Message&>(*original).GetRefCount(), 2);
+        EXPECT_EQ(static_cast<const shmem::Message&>(*copy).GetRefCount(), 2);
```
Comment (dennisklein) [inline test/message/_message.cxx:251]:
```suggestion
    if ("shmem" == transport) {
        EXPECT_EQ(static_cast<const shmem::Message&>(*original).GetRefCount(), 2);
        EXPECT_EQ(static_cast<const shmem::Message&>(*copy).GetRefCount(), 2);
    }
```

Comment (dennisklein) [reply test/message/_message.cxx:251]: alternatively, remove the transport function argument or assert it to be "shmem".

File: fairmq/Message.h [hunk 3]
```diff
@@ -47,6 +47,10 @@ struct Message
     TransportFactory* GetTransport() { return fTransport; }
     void SetTransport(TransportFactory* transport) { fTransport = transport; }
 
+    /// Copy the message buffer from another message
+    /// Transport may choose not to physically copy the buffer, but to share across the messages.
+    /// Modifying the buffer after a call to Copy() is undefined behaviour.
```
Comment (dennisklein) [inline fairmq/Message.h:52]: :+1:

File: fairmq/shmem/Manager.h [hunk 4]
```diff
@@ -71,7 +70,58 @@ struct ShmPtr
...
+        } else {
+            return nullptr;
+        }
+    }
+
+    static uint16_t RefCount(char* rPtr)
+    {
+        auto rcPtr = reinterpret_cast<std::atomic<uint16_t>*>(rPtr
+                                                              + sizeof(uint16_t) * 2
+                                                              + *(reinterpret_cast<uint16_t*>(rPtr + sizeof(uint16_t))));
```
Comment (dennisklein) [inline fairmq/shmem/Manager.h:84]: Could you please add some inline comments that explain what your are doing in those terms in english words? A general comment describing the byte layout of the `ShmHeader` would be really helpful, too. Thx!

**Review by dennisklein — CHANGES_REQUESTED** [review]

File: fairmq/shmem/Manager.h (hunk 4 above)
Comment (rbx) [reply fairmq/shmem/Manager.h:84]: I have the layout in a comment a bit further in the code, but sure I'll pull it up and add some prose.

**Review by dennisklein — APPROVED** [review]

**Pull request merged**